torch/utils/data/dataloader.py

```diff
@@ -371,19 +371,8 @@ def _shutdown_workers(self):
                 self.done_event.set()
                 for q in self.index_queues:
                     q.put(None)
-                # if some workers are waiting to put, make place for them
-                try:
-                    while not self.worker_result_queue.empty():
-                        self.worker_result_queue.get()
-                except (FileNotFoundError, ImportError):
-                    # Many weird errors can happen here due to Python
-                    # shutting down. These are more like obscure Python bugs.
-                    # FileNotFoundError can happen when we rebuild the fd
-                    # fetched from the queue but the socket is already closed
-                    # from the worker side.
-                    # ImportError can happen when the unpickler loads the
-                    # resource from `get`.
-                    pass
+                # Workers can't be waiting to put because they each has only at
+                # most 2 batches, which should be far below SimpleQueue's limit.
```

Comment:
No. The pipe has a certain size limit. Waiting to put means that the pipe already hit the limit but the work is still trying to put. But now each worker has its own queue, so the max number of batches in queue is 2, as long as each batch is less than half of the pipe capacity, it should be fine. Pipe capacity on mac & linux is typically 16KB or 64KB.

Reply:
I don't think that's a good idea. The pipe buffer size on my Ubuntu host is 1MB, and assuming that batches are smaller than 0.5MB is too fragile. People are loading large images and videos using those classes.

Reply:
@apaszke makes sense. Can I just `except Exception` here then? I've found more possible errors here due to the `.get`: `EOFError`, etc. It is hard to make it a comprehensive list.

Reply:
@apaszke I was adding more exceptions. However, I noticed that `.get` can be very slow here when workers have already exited. It needs some time before reporting `connection reset by peer`.  This can be troublesome when we exit in the middle of an epoch, i.e. `data_queue` might be non-empty. So I changed the `SimpleQueue` to `Queue` because `Queue`'s `put` is non-blocking, and added `cancel_join_thread` call in workers so that they don't wait for the putting thread when exiting. This exits cleanly from my local tests. Let me know what you think! Thanks!